--- a/networking/netmsg.c
+++ b/networking/netmsg.c
@@ -25,12 +25,21 @@ int netmsg_main(int argc, char **argv)
 	struct sockaddr_in addr;
 	int optval = 1;
 	unsigned char buf[1001];
+	char *msg = argv[2];
+	unsigned long port = 0x1337;
 
-	if (argc != 3) {
-		fprintf(stderr, "usage: %s <ip> \"<message>\"\n", argv[0]);
+	if (argc != 3 && argc != 4) {
+		fprintf(stderr, "usage: %s <ip> [port] \"<message>\"\n", argv[0]);
 		exit(1);
 	}
 
+	if (argc == 4) {
+		msg = argv[3];
+		port = strtoul(argv[2], NULL, 10);
+		if (port < 0)
+			port = 0x1337;
+	}
+
 	if ((s = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
 		perror("Opening socket");
 		exit(1);
@@ -39,13 +48,18 @@ int netmsg_main(int argc, char **argv)
 	memset(&addr, 0, sizeof(addr));
 	addr.sin_family = AF_INET;
 	addr.sin_addr.s_addr = inet_addr(argv[1]);
-	addr.sin_port = htons(0x1337);
+	addr.sin_port = htons(port);
 
 	memset(buf, 0, 1001);
+	if (argc == 4) {
+	strncpy(buf, msg, 998);
+	}
+	else {
 	buf[0] = 0xde;
 	buf[1] = 0xad;
 
-	strncpy(buf + 2, argv[2], 998);
+	strncpy(buf + 2, msg, 998);
+	}
 
 	if (setsockopt (s, SOL_SOCKET, SO_BROADCAST, (caddr_t) &optval, sizeof (optval)) < 0) {
 		perror("setsockopt()");
